---
title: "README"
author: "Ben Cortes"
date: "October 2, 2018"
output: html_document
---

Load tidyverse

```{r}
library(tidyverse)
```

Download fang_et_al_genotypes.txt and SNP_position.txt to the project folder and rename them

```{r}
download.file("https://raw.githubusercontent.com/EEOB-BioData/BCB546X-Fall2018/master/assignments/UNIX_Assignment/fang_et_al_genotypes.txt", destfile = "data/genotypes.txt")

download.file("https://raw.githubusercontent.com/EEOB-BioData/BCB546X-Fall2018/master/assignments/UNIX_Assignment/snp_position.txt", destfile = "data/SNPs.txt")
```
Inspect file structure
  file.info
    size - size of file in bytes
    isdir - whether or not file is directory
    mtime - last time file modified
    ctime - "last status change"
    atime - last accession time
    
```{r}
as.tibble(file.info("data/genotypes.txt"))[,c(1,2,4:6)]
as.tibble(file.info("data/SNPs.txt"))[,c(1,2,4:6)]
```
Inspect file structure
  Turn both text files into tibbles and call them
  Numbers of rows and columns per each file are written below the tibble
  Setting col_names equal to FALSE, circumvents the issues with the transpose command

```{r}
genotypes <- read_tsv("data/genotypes.txt", col_names = F)
snps <- read_tsv("data/SNPs.txt")

genotypes
snps
```

Get the three useful snp columns

```{r}
snps_f <- select(snps, 1,3,4)
```

Transpose the imported genotype.txt file
  Transpose will not transpose headers. Therefore, the header was incorporated into the 
    tibble by setting col_names equal to FALSE earlier
  Return first row to columns following transposition
  
```{r}
tran_genotypes <- as.tibble(t(genotypes))
#names(tran_genotypes) <- as.character(unlist(tran_genotypes[1,]))
#tran_genotypes <- tran_genotypes[-1,]
tran_genotypes
```

Pull maize and teosinte data from tran_genotypes
  Eliminate unneeded rows at top (JG_OTU and Group)
  Return top row to header

```{r}
maize_tgen <- as.tibble(select(tran_genotypes, grep("ZMM", tran_genotypes)))
teos_tgen <- as.tibble(select(tran_genotypes, grep("ZMP", tran_genotypes)))
maize_tgen <- maize_tgen[-c(2,3),]
teos_tgen <- teos_tgen[-c(2,3),]
names(maize_tgen) <- as.character(unlist(maize_tgen[1,]))
maize_tgen <- maize_tgen[-1,]
names(teos_tgen) <- as.character(unlist(teos_tgen[1,]))
teos_tgen <- teos_tgen[-1,]

```

Combine snps_f and each of the tgen files, forming maize and teos respectively
Arrange by Chromosome, both in ascending and descending order

```{r}
maize_asc <- arrange(as.tibble(cbind(snps_f, maize_tgen[,-1])), as.numeric(Chromosome), as.numeric(Position))
teos_asc <- arrange(as.tibble(cbind(snps_f, teos_tgen[,-1])), as.numeric(Chromosome), as.numeric(Position))
maize_des <- arrange(as.tibble(cbind(snps_f, maize_tgen[,-1])), as.numeric(Chromosome), desc(as.numeric(Position)))
teos_des <- arrange(as.tibble(cbind(snps_f, teos_tgen[,-1])), as.numeric(Chromosome), desc(as.numeric(Position)))
```

For the descending files, replace "?" in "?/?" with "-" in "-/-". For whatever reason, the it only works if it is done with lapply. teos_des <- as.tibble(gsub("\\?", "-", teos_des)) runs, but the entire tibble gets completely messed up with replacements. Not sure why.

```{r}
maize_des <- as.tibble(lapply(maize_des, function(x) {
  gsub("\\?", "-", x) 
}))
teos_des <- as.tibble(lapply(teos_des, function(x) {
  gsub("\\?", "-", x) 
}))

```

Separated files
  Create function to split rows of dataframe into smaller files. Files with identical values in a user-defined, ordered column will be added to same file (ie, all files with same value in a column will be placed into a file). Output to files with filenames bearing the pattern "'third input' 'value of column at given iteration' 'fourth input'". Input: dataframe, number of column, anything to be written in the output filename before the column value (including the file path), anything to be written in the output filename after the column value (including the file type).
  Split

```{r}
split_by_ordered_column <- function(fname, cnum, prefix, suffix) {

#dataframe to split = fname
#column to split by = cnum
#first part of filename = prefix
#second part of filename = suffix

for (i in 1:nrow(fname)) {
    if(i == 1 || fname[i,cnum] != fname[i-1,cnum]) {
        write_tsv(fname[i,], paste(prefix, fname[i,cnum], suffix, sep = ""))
    } else {
        write_tsv(fname[i,], paste(prefix, fname[i,cnum], suffix, sep = ""), append = T)
    }
}
  
}

split_by_ordered_column(maize_asc, 2, "sorted_data/chromosome_", "_maize_asc.txt")
split_by_ordered_column(teos_asc, 2, "sorted_data/chromosome_", "_teos_asc.txt")
split_by_ordered_column(maize_des, 2, "sorted_data/chromosome_", "_maize_des.txt")
split_by_ordered_column(teos_des, 2, "sorted_data/chromosome_", "_teos_des.txt")
```







